library/common/main_interface.cc

```diff
@@ -67,62 +51,7 @@ envoy_status_t set_preferred_network(envoy_network_t network) {
-  // here.
-  Envoy::Extensions::Clusters::DynamicForwardProxy::forceRegisterClusterFactory();
-  Envoy::Extensions::HttpFilters::DynamicForwardProxy::
-      forceRegisterDynamicForwardProxyFilterFactory();
-  Envoy::Extensions::HttpFilters::RouterFilter::forceRegisterRouterFilterConfig();
-  Envoy::Extensions::NetworkFilters::HttpConnectionManager::
-      forceRegisterHttpConnectionManagerFilterConfigFactory();
-  Envoy::Extensions::TransportSockets::RawBuffer::forceRegisterDownstreamRawBufferSocketFactory();
-  Envoy::Extensions::TransportSockets::RawBuffer::forceRegisterUpstreamRawBufferSocketFactory();
-  Envoy::Extensions::TransportSockets::Tls::forceRegisterUpstreamSslSocketFactory();
-  Envoy::Upstream::forceRegisterLogicalDnsClusterFactory();
-
-  // Initialize the server's main context under a try/catch loop and simply
-  // return EXIT_FAILURE as needed. Whatever code in the initialization path
-  // that fails is expected to log an error message so the user can diagnose.
-  // Note that in the Android examples logging will not be seen.
-  // This is a known problem, and will be addressed by:
-  // https://github.com/lyft/envoy-mobile/issues/34
-  try {
-    main_common_ = std::make_unique<Envoy::MainCommon>(5, envoy_argv);
-
-    // Note: init_engine must have been called prior to calling run_engine or the creation of the
-    // envoy runner thread.
-
-    // Note: We're waiting longer than we might otherwise to drain to the main thread's dispatcher.
-    // This is because we're not simply waiting for its availability and for it to have started, but
-    // also because we're waiting for clusters to have done their first attempt at DNS resolution.
-    // When we improve synchronous failure handling and/or move to dynamic forwarding, we only need
-    // to wait until the dispatcher is running (and can drain by enqueueing a drain callback on it,
-    // as we did previously).
-    stageone_callback_handler_ = main_common_->server()->lifecycleNotifier().registerCallback(
-        Envoy::Server::ServerLifecycleNotifier::Stage::PostInit, []() -> void {
-          http_dispatcher_->ready(main_common_->server()->dispatcher(),
-                                  main_common_->server()->clusterManager());
-        });
-  } catch (const Envoy::NoServingException& e) {
-    return ENVOY_SUCCESS;
-  } catch (const Envoy::MalformedArgvException& e) {
-    std::cerr << e.what() << std::endl;
-    return ENVOY_FAILURE;
-  } catch (const Envoy::EnvoyException& e) {
-    std::cerr << e.what() << std::endl;
-    return ENVOY_FAILURE;
-  }
-
-  // Run the server listener loop outside try/catch blocks, so that unexpected
-  // exceptions show up as a core-dumps for easier diagnostics.
-  return main_common_->run() ? ENVOY_SUCCESS : ENVOY_FAILURE;
+envoy_status_t run_engine(envoy_engine_t, const char* config, const char* log_level) {
+  engine_ = std::make_unique<Envoy::Engine>(config, log_level, preferred_network_);
```

Comment:
Might be good to link an issue for making Envoy instance-based rather than an `engine_` singleton here

Reply:
We do have an existing issue, and it's referenced above on L42. But we could reference it here as well.